### PR TITLE
Detect which reference examples should check/codegen/verify.

### LINKED ISSUE #348: Automatically detect which reference examples should check/codegen/verify.

### PR DESCRIPTION
**Pull request by bdalrhm**
### Description of changes: 

Currently, the dashboard fails says that RMC panics in 58 examples in the Rust Reference. However, that number is not accurate because the dashboard ignores configurations that many examples need to run. This PR maps those configurations to the closest options provided by compiletest and adds them as comments to the examples before running compiletest. This change reduces the number failing tests to 19.

### Resolved issues:

Resolves #348

### Testing:

* How is this change tested?
  * The dashboard, printed in the comment below, has fewer failing test now.
* Is this a refactor change?
  * A lot of the code was rewritten to account for the line number of code blocks.

### Checklist
- [x] Each commit message has a non-empty body, explaining why the change was made
- [x] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [x] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by bdalrhm** [comment]
The current dashboard (compare with [old dashboard](https://github.com/model-checking/rmc/pull/325#issuecomment-879924820https://github.com/model-checking/rmc/pull/325#issuecomment-879924820)):
```
$ ./x.py run -i --stage 1 dashboard
Updating only changed submodules
Submodules updated in 0.02 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
Generating confidence dashboard
Building stage0 tool dashboard (x86_64-unknown-linux-gnu)
    Finished release [optimized + debuginfo] target(s) in 0.13s
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
    Finished release [optimized + debuginfo] target(s) in 0.12s
warning: unused import: `CodegenResults`
 --> compiler/rustc_codegen_ssa/src/traits/backend.rs:4:13
  |
4 | use crate::{CodegenResults, ModuleCodegen};
  |             ^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: 1 warning emitted

warning: unused import: `rustc_middle::ty::subst::InternalSubsts`
  --> compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs:13:5
   |
13 | use rustc_middle::ty::subst::InternalSubsts;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[warn(unused_imports)]` on by default

warning: unused imports: `Binder`, `GenericParamDefKind`, `TraitRef`
  --> compiler/rustc_codegen_llvm/src/gotoc/rvalue.rs:15:11
   |
15 |     self, Binder, GenericParamDefKind, Instance, IntTy, TraitRef, Ty, UintTy, VtblEntry,
   |           ^^^^^^  ^^^^^^^^^^^^^^^^^^^                   ^^^^^^^^

warning: function is never used: `time_trace_profiler_finish`
   --> compiler/rustc_codegen_llvm/src/llvm_util.rs:151:8
    |
151 | pub fn time_trace_profiler_finish(file_name: &str) {
    |        ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: 3 warnings emitted

warning: associated function is never used: `process_rlink`
   --> compiler/rustc_driver/src/lib.rs:596:8
    |
596 |     fn process_rlink(sess: &Session, compiler: &interface::Compiler) -> Result<(), ErrorReported> {
    |        ^^^^^^^^^^^^^
    |
    = note: `#[warn(dead_code)]` on by default

warning: 1 warning emitted

    Finished release [optimized + debuginfo] target(s) in 0.18s
    Finished release [optimized + debuginfo] target(s) in 0.12s
    Finished release [optimized + debuginfo] target(s) in 0.13s
Some tests failed in compiletest suite=ref mode=rmc host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
# of tests: 322 ✔️ 303   ❌ 19
ref ✔️ 303 ❌ 19
  Appendices ✔️ 1 ❌ 1
    Macro Follow-Set Ambiguity Formal Specification ✔️ 1
    Glossary ❌ 1
  Attributes ✔️ 27 ❌ 2
    Diagnostics ✔️ 13
    Code generation ✔️ 5
    Derive ✔️ 2
    Limits ✔️ 2 ❌ 1
    Testing ✔️ 3
    Type System ✔️ 1
  Application Binary Interface ✔️ 3
  Conditional compilation ✔️ 2
  Crates and source files ✔️ 1 ❌ 1
  Items ✔️ 95 ❌ 2
    Associated Items ✔️ 11
    Enumerations ✔️ 8
    Constant items ✔️ 4
    External blocks ✔️ 3
    Functions ✔️ 16
    Generic parameters ✔️ 9
    Implementations ✔️ 7
    Traits ✔️ 15
    Modules ✔️ 1
    Structs ✔️ 4
    Static items ✔️ 2
    Type aliases ✔️ 2
    Unions ✔️ 8
    Use declarations ✔️ 5 ❌ 2
  Introduction ✔️ 1
  Lexical structure ✔️ 9 ❌ 1
    Keywords ❌ 1
    Tokens ✔️ 8
    Comments ✔️ 1
  Macros ✔️ 10 ❌ 1
    Macros By Example ✔️ 10
  Linkage ✔️ 1 ❌ 1
  Names ✔️ 19
    Namespaces ✔️ 1
    Paths ✔️ 13
    Preludes ✔️ 1
    Visibility and privacy ✔️ 4
  Memory model ✔️ 1
    Variables ✔️ 1
  Patterns ✔️ 21 ❌ 1
  Statements and expressions ✔️ 58 ❌ 7
    Expressions ✔️ 55 ❌ 7
      Block expressions ✔️ 3 ❌ 2
      Array and index expressions ✔️ 2
      Call expressions ✔️ 2
      Field access expressions ✔️ 2
      If and if let expressions ✔️ 4
      Grouped expressions ✔️ 1 ❌ 1
      Literal expressions ✔️ 1
      Loop expressions ✔️ 7 ❌ 1
      Match expressions ✔️ 4
      Method call expressions ✔️ 1 ❌ 1
      Operator expressions ✔️ 13 ❌ 2
      Path expressions ✔️ 1
      Range expressions ✔️ 3
      Return expressions ✔️ 1
      Struct expressions ✔️ 6
      Closure expressions ✔️ 1
      Tuple and index expressions ✔️ 1
    Statements ✔️ 3
  The Rust runtime ✔️ 1
  Type system ✔️ 53 ❌ 2
    Lifetime elision ✔️ 8
    Destructors ✔️ 8
    Subtyping and Variance ✔️ 3
    Trait and lifetime bounds ✔️ 5
    Type coercions ✔️ 6
    Type layout ✔️ 7
    Types ✔️ 16 ❌ 2
      Boolean type ✔️ 1
      Function item types ✔️ 2
      Closure types ✔️ 3 ❌ 1
      Impl trait type ✔️ 3
      Function pointer types ✔️ 1
      Type parameters ✔️ 1
      Array types ✔️ 1
      Slice types ✔️ 1
      Trait object types ❌ 1
      Inferred type ✔️ 1

Build completed successfully in 0:00:18
```

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -2461,6 +2465,13 @@ impl<'test> TestCx<'test> {
         // 2. It may pass some options that do not make sense for RMC
         // So we create our own command to execute RMC and pass it to self.compose_and_run_compiler(...) directly.
         let mut rmc = Command::new("rmc");
+        // We cannot pass rustc flags directly to RMC. Instead, we add them
+        // to the current enviornment through the `RUSTFLAGS` environment
```
Comment (nchong-at-aws) [inline src/tools/compiletest/src/runtest.rs:2469]: typo: environment

File: src/tools/dashboard/src/reference.rs [hunk 2]
```diff
@@ -54,87 +60,129 @@ fn parse_hierarchy(summary_path: &Path) -> HashMap<PathBuf, PathBuf> {
...
-fn preprocess_examples(_paths: Vec<&PathBuf>) {
-    // For now, we will only pre-process the tests that cause infinite loops.
-    // TODO: properly implement this step (see issue #324).
+/// Pre-processes the examples in `map` before running them with `compiletest`.
+fn preprocess_examples(map: &HashMap<Example, PathBuf>) {
+    // Copy compiler configurations specified in the original markdown code
+    // block.
+    for (from, to) in map.iter() {
+        let file = File::open(&from.path).unwrap();
+        let mut it = BufReader::new(file).lines().skip(from.line - 1);
```
Comment (nchong-at-aws) [inline src/tools/dashboard/src/reference.rs:167]: Why the -1? Add a comment please.

**Review by nchong-at-aws — CHANGES_REQUESTED** [review]

File: src/tools/dashboard/src/reference.rs (hunk 2 above)
Comment (bdalrhm) [reply src/tools/dashboard/src/reference.rs]: Done! Please take a look.

**Review by nchong-at-aws — APPROVED** [review]
Looks good. This PR adds heuristics to the preprocessing step of examples to map them to an expected outcome.

**Pull request merged**